### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by drnimbusrain**
This PR addresses issue #23 , and reorganizes the canopy_driver to be more streamlined and easier for future development.  Numerous new routines are added to handle the different main components of the Canopy-App, and has removed significant amounts of code that are not needed and at times redundant.

### CODE REVIEW

**Review by zmoon — COMMENTED** [review]
I'll look more later, but I think the subroutines that are in files by themselves would be better off grouped in a module, like `canopy_standalone_mod` or something. Or `canopy_driver_mod` and the main program could be `canopy_app` (which would be fun given the repo name).

**Comment by drnimbusrain** [comment]
Oh, I like it Zach, I can make it like that during your review.   I hope we
are approaching a better framework with each step.  :)

**Comment by zmoon** [comment]
I think d89f760 was maybe a step too far. I was envisioning that the WAF, phot, etc. routines, which might be called by hypothetical future CCPP or what-have-you driver, could stay separate. But routines that are specific to the standalone driver, could go the in the `canopy_driver_mod`. I'll ponder this more, though.

**Comment by drnimbusrain** [comment]
@zmoon OK, added WAF as separate option as discussed.  Will now revert and  pull out all major canopy options, canopy_wind, canopy_waf, canopy_phot, and canopy_eddy as separate submodules again to maintain modularity.

**Comment by drnimbusrain** [comment]
@zmoon Does this PR look OK now, or any additional comments?  After we merge this PR on the overall framework, we can split off using different branches for other features (including NetCDF support) and new modules. 
Thank you so much!

**Review by zmoon — COMMENTED** [review]
Some various comments, but main one is that all subroutines should be in modules, and imports in the main program should be explicit. I think most of the subroutines currently in their own file could go in the driver mod for now.

This is what the current program and module `use` connections look like [^1]. You can see that it doesn't tell the whole story, due to the above.

```mermaid
graph TD
	canopy_app
	canopy_app --> canopy_files_mod
	canopy_canmet_mod
	canopy_canopts_mod
	canopy_canvars_mod
	canopy_const_mod
	canopy_coord_mod
	canopy_driver_mod
	canopy_driver_mod --> canopy_const_mod
	canopy_driver_mod --> canopy_utils_mod
	canopy_eddy_mod
	canopy_eddy_mod --> canopy_const_mod
	canopy_files_mod
	canopy_phot_mod
	canopy_phot_mod --> canopy_const_mod
	canopy_txt_io_mod
	canopy_txt_io_mod --> canopy_canmet_mod
	canopy_txt_io_mod --> canopy_canopts_mod
	canopy_txt_io_mod --> canopy_canvars_mod
	canopy_txt_io_mod --> canopy_coord_mod
	canopy_utils_mod
	canopy_utils_mod --> canopy_const_mod
	canopy_waf_mod
	canopy_waf_mod --> canopy_const_mod
	canopy_waf_mod --> canopy_utils_mod
	canopy_wind_mod
	canopy_wind_mod --> canopy_const_mod
```

[^1]: Using [`fortdepend -g`](https://github.com/ZedThree/fort_depend.py) followed by `sed 's/ ->/ -->/g' canopy-app.dot | sed '1 s/^.*/```mermaid\ngraph TD/' | sed '$ s/^.*/```/'`

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90:80]: @zmoon From my understanding, the _rks are not needed here anymore and are redundant (causes warning in the debug compile), since these namelist variables are defined as real(rk) now in canopy_coord_mod.F90 ad canopy_vars_mod.F90.

**Comment by drnimbusrain** [comment]
@zmoon Thank you for the excellent review!  

For your main comment, I am a bit confused on how you suggest it should look though.  Can you derive a similar flow chart that describes what it _should_ look like, and/or would you mind making those suggested changes in another branch and doing a PR so I can better see how you envision the changes should look?   I am somewhat wondering why modules are better here than their own subroutines that house the read, alloc, init, and dealloc steps.   It seems this way keeps it compact in the main canopy-app program.  

The canopy-driver mod, in my mind, should contain the main driving components that describe the canopy, and should not necessarily contain the  read, alloc, init, and dealloc  as you suggest should be in that module.  

The main component options, which I agree should stay modular, are in modules I agree.  Thanks!

File: canopy_utils_mod.F90
Comment (drnimbusrain) [reply canopy_utils_mod.F90:90]: @zmoon   This comes from the definition of a spherical earth and the haversine formula for the distance between two points.  See:  https://en.wikipedia.org/wiki/Great-circle_distance

File: README.md
Comment (drnimbusrain) [reply README.md:45]: @zmoon The "testrun" is just an example output file prefix provided by the user.

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90:1]: @zmoon Not sure on this, see my main comment/response.

File: canopy_canopts_mod.F90
Comment (drnimbusrain) [reply canopy_canopts_mod.F90:12]: @zmoon Nice catch, and I agree.  I have since removed in latest commit.

File: canopy_app.F90
Comment (drnimbusrain) [reply canopy_app.F90:9]: @zmoon Yes, I see your point.  But is that fully necessary, as the canopy_app code is very compact now as I have chosen to employ the major parts of the codes as their own subroutines files (I know you don't like), which keeping the actual canopy physical calculations as modules (e.g., wind, WAF, phot, and eddy)?

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90:1]: @zmoon I see your point, but may consider this to be standalone subroutine as major part of canopy-app. See my comments above.

File: canopy_read_txt.F90
Comment (drnimbusrain) [reply canopy_read_txt.F90:2]: @zmoon  Please see my response above to canopy_write_txt.  I also envisioned having a canopy_read_ncf.F90 canopy_write_ncf.F90.

File: canopy_dealloc.F90
Comment (drnimbusrain) [reply canopy_dealloc.F90:2]: @zmoon I see your point, but may consider this to be standalone subroutine as major part of canopy-app. See my comments above.

File: canopy_calcs.F90
Comment (drnimbusrain) [reply canopy_calcs.F90:2]: @zmoon I see your point, but may consider this to be standalone subroutine as major part of canopy-app. See my comments above.

File: canopy_alloc.F90
Comment (drnimbusrain) [reply canopy_alloc.F90:2]: @zmoon I see your point, but may consider this to be standalone subroutine as major part of canopy-app. See my comments above.

File: canopy_readnml.F90
Comment (zmoon) [reply canopy_readnml.F90:80]: It seems the error was because it was defined in those multiple modules that were being used, so it was an "ambiguous reference". With your change it should be ok to `use canopy_const_mod, only: rk` and put them back.

Just because it is declared as `real(rk)` doesn't mean this doesn't have an effect, here is an example:
```fortran
program asdf
  use iso_fortran_env, only: rk => real64
  implicit none

  real(rk) :: x

  print *, 0.123
  print *, 0.123_rk

  x = 0.123
  print *, x
  x = 0.123_rk
  print *, x

end program
```
```
  0.123000003
  0.12300000000000000
  0.12300000339746475
  0.12300000000000000
```
Note the third one, from `x = 0.123`.

File: canopy_canopts_mod.F90
Comment (zmoon) [reply canopy_canopts_mod.F90]: 👍

File: README.md
Comment (zmoon) [reply README.md]: Ok, I think it makes more sense now

File: canopy_utils_mod.F90
Comment (zmoon) [reply canopy_utils_mod.F90:90]: The link you shared notes that the formula you are using is from the spherical law of cosines, and that it doesn't work well for small distances, recommending to use a haversine formula instead. Also they are recommending 6371 km for Earth radius.

I'm not saying this function won't be useful. But to me `dx` implies latitude circle, which is a simpler problem.

File: canopy_app.F90
Comment (zmoon) [reply canopy_app.F90:9]: I support keeping the canopy-app program compact, I think that is nice. But you get the same benefit (reusability and flexibility) when you put those subroutines in modules and `use` them.

And for potential other applications, we want to be providing a library with friendly and well-defined API (module or modules) that the client can `use`.

File: canopy_readnml.F90
Comment (zmoon) [reply canopy_readnml.F90:1]: > may consider this to be standalone subroutine as major part of canopy-app

Could you expand on this a bit more for me?

File: canopy_utils_mod.F90
Comment (drnimbusrain) [reply canopy_utils_mod.F90:90]: I am having trouble finding the exact reference, but I think this formula is indeed a haversine formula.

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90:80]: @zmoon Thank you.  I will put them back in those modules that had the ambiguous references.

File: canopy_app.F90
Comment (drnimbusrain) [reply canopy_app.F90:9]: @zmoon OK, I defer to you if this is better practice.  It would be awesome if you can help make the changes from another branch and do a PR.

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90:1]: @zmoon Maybe this style is an issue, I agree with orphan subroutines, but to me this keeps the canopy-app program more compact.  Maybe not.  I definitely appreciate any help you can provide here on the coding.

File: canopy_readnml.F90
Comment (drnimbusrain) [reply canopy_readnml.F90]: @zmoon I have added them back in for the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/30/commits/c8c8bd95bc88bfa3ef6102c1551a48757536b46c).   I think this was the only place I removed them, but if you see others I appreciate it!  Thanks!

File: canopy_utils_mod.F90
Comment (zmoon) [reply canopy_utils_mod.F90:90]: A haversine formula would have some angles divided by 2 in it ( $\mathrm{hav}(θ) = \sin^2(θ/2)$ ). This is shown on that page you sent. They also point out a third formula that is supposed to be better than the haversine one.

File: canopy_app.F90
Comment (zmoon) [reply canopy_app.F90:9]: Yes happy to have a go at that, and we can discuss tomorrow.

For reference, another issue with subroutines outside of modules is that the compiler [^b] isn't able to check for you that you are passing the correct types [^a]. which can lead to obscure bugs.

```fortran
subroutine s(x)
  real, intent(in) :: x
  print *, 2 * x
end subroutine s
```
```fortran
program p
  call s(2.)
  call s(2)
end program p
```
```
> gfortran-9 s.f90 p.f90 && ./a.out
   4.00000000
   5.60519386E-45
```

[^a]: Unless, in the place you are using it, you define an explicit interface using an interface block
[^b]: At least GFortran 9. 10+ seem to be ok, but maybe not in all situations?

File: canopy_utils_mod.F90
Comment (drnimbusrain) [reply canopy_utils_mod.F90:90]: @zmoon OK, thanks.  Please if you want to modify and add a PR in different branch, please do so.   This seems to work and get a good estimate on the distance, but of course open to any changes.

Comment (drnimbusrain) [reply canopy_utils_mod.F90:90]: @zmoon OK, I have just compared the formula I have in there, and your recommended(R * cos(lat) * (lon2 - lon1) / 180 * pi)

**Single grid cell example:**
Current:
8527.1568277502665 m
Recommended:
8518.0255659081304

Pretty close, but I will make the change you suggest in next commit.

Comment (zmoon) [reply canopy_utils_mod.F90]: Interesting, thanks for checking out the formulas.

If you want to be more precise with the 6371 km, it looks like they have 6371.0088 km in their figure. And for reference in the code, could say that this is the WGS84 arithmetic mean radius.

File: canopy_const_mod.F90
Comment (zmoon) [inline canopy_const_mod.F90:40]:
```suggestion
    REAL(RK),          PARAMETER     :: rearth = 6371008.8_rk  ! WGS84 arithmetic mean radius
```

File: canopy_utils_mod.F90
Comment (zmoon) [reply canopy_utils_mod.F90]: I did some investigation of the other formulas. The current formula does suffer bit at very small distances, as noted in the Wikipedia article, but probably wouldn't matter.

However, more importantly, I don't think great circle distance is what we want here. A great circle's plane always passes through the Earth's center. So even when we use lat1=lat2 in the formulas, the paths taken are not exactly along latitude circles, they curve (i.e., the latitude of the path is not constant). For grid dx, we (usually) want the distance along a latitude circle (constant latitude), which the formula I gave is exact for.

<details><summary>Some figs:</summary>

great-circle-dx_fig1
great-circle-dx_fig2
great-circle-dx_fig3

</details>

File: canopy_readnml.F90
Comment (zmoon) [reply canopy_readnml.F90]: If I notice others I will comment, but will now resolve this one. Thanks!

File: canopy_read_txt.F90
Comment (zmoon) [reply canopy_read_txt.F90:2]: We'll keep this in mind. Another idea could be to have an io module that has text and nc submodules

File: canopy_utils_mod.F90
Comment (zmoon) [reply canopy_utils_mod.F90]: I will try with the simple formula, and check to make sure not seeing negatives.

File: canopy_app.F90
Comment (zmoon) [reply canopy_app.F90:9]: Will be working on this in a PR.

File: canopy_readnml.F90
Comment (zmoon) [reply canopy_readnml.F90:1]: Will be working on this on a PR.

**Comment by zmoon** [comment]
@drnimbusrain I changed the dx calc. Check out [3ea35e1](https://github.com/noaa-oar-arl/canopy-app/pull/30/commits/3ea35e1eb2649cdea915433b97d42cd225a33248) if you want to see the comparison. They were same to 6 or 7 sig figs I think. I put abs on the dlon so that it won't be negative, but I also confirmed.

**Comment by drnimbusrain** [comment]
@zmoon Yes, I like your dx change and have tested it.  Can you approve this PR so we can merge now and branch off?

**Review by zmoon — APPROVED** [review]

**Pull request merged**
